design_docs/openapi.yml

```diff
@@ -0,0 +1,162 @@
+      parameters:
+        - name: stream
+          in: query
+          schema:
+            type: boolean
+            default: false # TODO: consider if `true` (stream by default) is better
+          description: |
+            Whether or not the server should stream responses
+            or wait for all responses to be generated before sending them back
+            to the client.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                $ref: "#/components/schemas/Message"
+        required: true
+      responses:
+        "200":
+          description: |
+            Successful operation. Respond with text.
+            Response format depends on `stream` query parameter.
+          content:
+            text/event-stream:
+              schema:
+                type: string
+            text/plain:
+              schema:
+                type: string
+        "400":
+          description: Invalid input
+        "401":
+          description: Unauthorized
+        "500":
+          description: Internal server error
+      security:
+        - CORS: []
+
+  /rate:
+    post:
+      operationId: rate
+      summary: Rate a response
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/RateRequest"
+      responses:
+        "204":
+          description: Successful operation
+        "400":
+          description: Invalid input
+        "401":
+          description: Unauthorized
+        "404":
+          description: Not found, data with URL does not exist
+        "500":
+          description: Internal server error
+      security:
+        - CORS: []
+
+components:
+  securitySchemes:
+    CORS: # TODO: validate that this is the correct way to represent CORS as security schema in OpenAPI spec
+      type: apiKey
+      in: header
+      name: Origin
+    ApiKeyAuth:
+      type: apiKey
+      in: header
+      name: X-API-KEY
+  schemas:
+    Message:
+      type: object
+      properties:
+        role:
+          type: string
+          enum: [assistant, user]
+        content:
+          type: string
+          format: markdown
+          description: |
+            Markdown-formatted response to user's chat message in the context
+            of the conversation.
+        timestamp:
+          type: integer
+          description: |
+            Unix timestamp for when the message was sent. Only present when message
+            is in the response.
+      required:
+        - role
+        - content
+    RateRequest:
+      type: object
+      properties:
+        messages:
+          type: array
+          items:
+            $ref: "#/components/schemas/Message"
+        rating:
+          type: boolean
+          description: Set to `true` if the user liked the response, `false` otherwise.
+      required: [url, rating]
```

Comment:
```suggestion
    RateRequest:
      type: object
      properties:
        messages:
          type: array
          items:
            $ref: "#/components/schemas/Message"
        rating:
          type: boolean
          description: Set to `true` if the user liked the response, `false` otherwise.
        id:
          type: string
          description: conversation ID
        ip_address:
          type: string
          description: user IP address
        message_index: 
          type: number
          description: Index in the messages array of message being updated.
      required: [url, rating, id, ip_address, message_index]
```